src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/ISolutionExtensions.cs

```diff
@@ -52,17 +52,32 @@ public static Document GetRequiredDocument(this Solution solution, DocumentId do
             throw new ArgumentNullException(nameof(documentId));
 
 #if !CODE_STYLE
-        // If we get a source-generated DocumentId, we can give a different exception to make it clear the type of failure this is; otherwise a failure of
-        // this in the wild is hard to guess whether this is because of a logic bug in the feature (where it tried to use a DocumentId for a document that disappeared)
-        // or whether it hasn't been correctly updated to handle source generated files.
         if (documentId.IsSourceGenerated)
+        {
+            // If we get a source-generated DocumentId, we can give a different exception to make it clear the type of failure this is; otherwise a failure of
+            // this in the wild is hard to guess whether this is because of a logic bug in the feature (where it tried to use a DocumentId for a document that disappeared)
+            // or whether it hasn't been correctly updated to handle source generated files.
             throw new ArgumentException($"{nameof(GetRequiredDocument)} was given a source-generated DocumentId, but it will never return a source generated document. The caller needs to be calling some other method.");
+        }
 #endif
 
         return solution.GetDocument(documentId) ?? throw CreateDocumentNotFoundException();
     }
 
 #if !CODE_STYLE
+    public static SourceGeneratedDocument GetRequiredSourceGeneratedDocumentForAlreadyGeneratedId(this Solution solution, DocumentId documentId)
```

Comment:
i don't mind this (but i want us all 3 to be ok here).  i do think this should be use din a few places as we have calls to TryGetSourceGeneratedDocumentForAlreadyGeneratedId that then throw.

i do want us to limit TryGetSourceGeneratedDocumentForAlreadyGeneratedId, and potentially remove if not needed.  i don't love that something could effectively change behavior based on the return value.  I *am* ok with the case where a client can say: in my scenario this must exists, so give it to me or crash.

Reply:
Doced and used more